### PR TITLE
Add ability to return all fields exported by a factory

### PR DESCRIPTION
**Pull request by mstemm**
Add the ability to return all fields exported by a factory. This is
important for programs like falco that need to validate rule filter
expressions for various event sources, as well as print out sets of
supported fields.

Previously, falco did direct calls to
sinsp::get_filtercheck_fields_info but we're trying to standardize
everything to work through factories, to make it easier to support new
event sources. This PR supports that work.

Signed-off-by: Mark Stemm <mark.stemm@gmail.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-ebpf

> /area libscap

/area libsinsp

> /area tests

> /area proposals

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new: add ability to return all fields exported by a filter factory. This will be used by Falco to support multiple event sources.
```

### CODE REVIEW

File: userspace/libsinsp/gen_filter.h
Comment (gnosek) [inline userspace/libsinsp/gen_filter.h:237]: Why lists and not vectors?

File: userspace/libsinsp/filter.h
Comment (gnosek) [inline userspace/libsinsp/filter.h:217]: I assume this is the only subclass?

File: userspace/libsinsp/filter.cpp
Comment (gnosek) [inline userspace/libsinsp/filter.cpp:2428]: range-based for?

File: userspace/libsinsp/filter.cpp
Comment (gnosek) [inline userspace/libsinsp/filter.cpp:2444]: fci->m_fields isn't a vector we could iterate over without an explicit index, is it?

File: userspace/libsinsp/filter.cpp
Comment (gnosek) [inline userspace/libsinsp/filter.cpp:2450]: What do these flags mean and why do we want this check here?

**Review by gnosek — COMMENTED** [review]
Hmm okay 🤷‍♂️ 

There's no user of this code added so I have ~zero context on what this is supposed to do

File: userspace/libsinsp/gen_filter.h
Comment (mstemm) [reply userspace/libsinsp/gen_filter.h:237]: No need for random access, just iteration.

File: userspace/libsinsp/filter.h
Comment (mstemm) [reply userspace/libsinsp/filter.h:217]: In the libraries, yes. In falco there's another subclass for "json_events", which are used by the k8s audit rules: https://github.com/leogr/falco/blob/new/plugin-system-api-additions/userspace/engine/json_evt.h#L376.

And after these changes are all merged, we'll be creating a subclass for plugins, so a given plugin can export the set of valid fields used by that plugin.

File: userspace/libsinsp/filter.cpp
Comment (mstemm) [reply userspace/libsinsp/filter.cpp:2428]: Sure, updated.

File: userspace/libsinsp/filter.cpp
Comment (mstemm) [reply userspace/libsinsp/filter.cpp:2444]: that's right, looks like it's been a plain array the whole time: https://github.com/falcosecurity/libs/blob/factory-add-get-fields/userspace/libsinsp/sinsp.h#L143

File: userspace/libsinsp/filter.cpp
Comment (mstemm) [reply userspace/libsinsp/filter.cpp:2450]: They are defined here: https://github.com/falcosecurity/libs/blob/factory-add-get-fields/userspace/libsinsp/event.h#L44. Generally if a field can't be used to filter events, we don't want to print it.

(And btw, this is just moving code from falco that did this in a specific way for syscall filterchecks. This PR makes it generic for a "factory" e.g. source of filters/filterchecks/fields).

**Comment by mstemm** [comment]
By the way, all of these PRs are used in this Falco PR (https://github.com/falcosecurity/falco/pull/1715), if you want to see how they are used.

File: userspace/libsinsp/gen_filter.h
Comment (gnosek) [reply userspace/libsinsp/gen_filter.h:237]: Still, I don't think a list should be the default choice for a container, due to the horrible memory accesses. Probably doesn't matter much here, but maybe at least we'd get cheap .size() with a vector

File: userspace/libsinsp/filter.cpp
Comment (gnosek) [reply userspace/libsinsp/filter.cpp:2442]: :sadpanda: might be nice to convert it to a vector later.

File: userspace/libsinsp/filter.cpp
Comment (gnosek) [reply userspace/libsinsp/filter.cpp:2448]: Thanks for the explanation.

> if a field can't be used to filter events, we don't want to print it

this should live as a comment above this `if`

File: userspace/libsinsp/gen_filter.h
Comment (mstemm) [reply userspace/libsinsp/gen_filter.h:237]: According to https://stackoverflow.com/questions/228908/is-listsize-really-on .size() is supposed to be O(1) and actually is in gcc >= 5.0. Also, the only use of this method just iterates (once) anyway.

File: userspace/libsinsp/filter.cpp
Comment (mstemm) [reply userspace/libsinsp/filter.cpp:2452]: Okay, added.

**Review by leogr — APPROVED** [review]
/approve

LGTM, just a note since the next Falco release is very close.

The new `get_fields()` conflicts with the one already present [here in Falco](https://github.com/leogr/falco/blob/new/plugin-system-api-additions/userspace/engine/json_evt.h#L389) (so basically Falco won't compile until the signature gets adjusted in Falco). 
I know that's addressed by https://github.com/falcosecurity/falco/pull/1715. However, we likely have to upgrade the version of the libs in Falco soon and I'm not confident https://github.com/falcosecurity/falco/pull/1715 will get merged in time. 

Thus, I wouldn't risk blocking the Falco release. So,
/hold
for the moment.

**Comment by mstemm** [comment]
That's fine--I wouldn't merge any of #74, #75, #76, or #77 individually. Just merge them all at once (after the release, most likely) and then I can fix things up with the corresponding falco release.

**Comment by leogr** [comment]
> That's fine--I wouldn't merge any of #74, #75, #76, or #77 individually. Just merge them all at once (after the release, most likely) and then I can fix things up with the corresponding falco release.

Perfect! I will review the other PRs soon.
Thank you.

**Review by gnosek — APPROVED** [review]

**Comment by mstemm** [comment]
Falco 0.30.0 is released so I'm removing the hold.
/hold cancel

**Review by gnosek — APPROVED** [review]

**Review by leogr — APPROVED** [review]
/approve

**Pull request merged**
